UI Redesigned

mojahid2021 opened the pull request:
Redesigned this app UI

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added Recharge/top-up entry, route-based Fare Calculator with station selection & swap, Recent Activity and Transactions summary, and expanded More screen.

* **UI/UX**
  * Major redesigns across Balance card, Fare Calculator, History, Transaction List, Home and More screens with animations, responsive cards, modern headers, and improved loading/empty/error states.

* **Localization**
  * Large English and Bengali string additions and refinements, including NFC/setup guidance and travel tips.

* **Chores**
  * App version bumped to 0.0.27; build tooling and Gradle wrapper updated.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

mojahid2021 committed:
feat: modernize UI for fare calculator and home screen

This commit introduces a significant UI refresh for the Fare Calculator and Home screens, adopting a more modern and visually appealing design.

**Key changes include:**

*   **Fare Calculator Screen:**
    *   Enhanced card-based layout for station selection and fare display.
    *   Improved visual hierarchy and information presentation.
    *   Modernized input fields and dropdowns with icons and refined styling.
    *   Gradient backgrounds and subtle elevations for a more polished look.
    *   Added a "Quick Tips" info card for better user experience.
*   **Home Screen (MainScreen):**
    *   Revamped layout with a welcome section, quick stats, and enhanced balance card.
    *   Modernized navigation bar with updated icons and selection indicators.
    *   Improved presentation of recent transaction history with a card-based, scrollable list.
    *   Consistent use of MaterialTheme colors and typography.
    *   Background gradients for a softer, more contemporary feel.
*   **General UI Enhancements:**
    *   Increased use of `Card` composables for better structure and elevation.
    *   Rounded corners and refined spacing for a cleaner aesthetic.
    *   Improved typography and color usage for better readability and visual appeal.
    *   Added previews for easier UI development and review.

Overall, these changes aim to provide a more intuitive, engaging, and modern user experience.

mojahid2021 committed:
feat: enhance UI for Fare Calculator

This commit introduces a significantly redesigned UI for the Fare Calculator screen, focusing on improved aesthetics, user experience, and information presentation.

Key changes:

- **Modernized Layout & Components:**
    - Updated card styles with rounded corners and subtle elevations.
    - Enhanced headers with icons and descriptive text for better context.
    - Improved station selection dropdowns with visual cues (icons, progress indicators).
    - Revamped fare display card with clearer information hierarchy and contextual styling.
    - Added animated transitions for a smoother user experience.
- **New Information Sections:**
    - Introduced a "Travel Information" card displaying route details, estimated time, distance, and potential savings with an MRT card.
    - Added a "Quick Tips" card offering helpful advice for commuters.
- **Improved Empty States:**
    - Redesigned empty state for the fare display, providing clearer instructions.
- **Visual Enhancements:**
    - Utilized Material Design icons more effectively.
    - Incorporated subtle background gradients.
    - Improved visual feedback for selected stations and card balance status.
    - Animated swap button for station selection.
- **Code Refinements:**
    - Extracted `StationDropdownField` into a reusable component.
    - Added `TravelInfoCard` and `QuickTipsCard` components.
    - Introduced `InfoRow` and `TipItem` helper composables for consistent styling.
- **Dependency Update:**
    - Added `compose.materialIconsExtended` for a wider range of icons.

These changes aim to make the Fare Calculator more intuitive, visually appealing, and informative for users.

mojahid2021 committed:
refactor(FareCalculator): improve animations and minor UI tweaks

- Optimized animation specifications for smoother transitions.
- Used `item(key = ...)` for LazyColumn items to improve performance and state preservation during recompositions.
- Set "Estimated Time" and "Distance" in TravelInfoCard to "N/A" as these features are not yet implemented.
- Removed unused progress indicator from `StationSelectionSection`.
- Used `Icons.AutoMirrored.Filled.TrendingUp` for better RTL support.

mojahid2021 committed:
refactor(FareCalculator): improve animations and minor UI tweaks

- Optimized animation specifications for smoother transitions.
- Used `item(key = ...)` for LazyColumn items to improve performance and state preservation during recompositions.
- Set "Estimated Time" and "Distance" in TravelInfoCard to "N/A" as these features are not yet implemented.
- Removed unused progress indicator from `StationSelectionSection`.
- Used `Icons.AutoMirrored.Filled.TrendingUp` for better RTL support.

mojahid2021 committed:
style: make background consistent throughout the app

Updated the background color of the FareCalculatorScreen and related components to use the theme's background color.
This ensures a consistent look and feel throughout the app.
Removed the gradient background and replaced it with a solid background color.

mojahid2021 committed:
chore: update version code and name to 0.0.24

mojahid2021 committed:
feat: add transparent color and enhance UI components in Fare Calculator

mojahid2021 committed:
refactor(FareCalculator): update layout and enhance transparency in UI components

mojahid2021 committed:
feat(BalanceCard, MainScreen, TransactionListScreen): enhance UI with new icons and improved layout

mojahid2021 commented:
![Screenshot_20250730_205912_MRT Buddy.jpg](https://github.com/user-attachments/assets/6d6bcf3d-ebe9-4c47-8cb5-892a5174bce4)

Everything is working fine in my device. Share your device model and crashing log

CodeByTarek commented:
@mojahid2021 I was checking the code on Pixel 5. I got 2 cards, one is Rapid Pass, the other one is MRT.  Works well for Rapid Pass, but the app crashed for MRT. Let me check the crash log or another phone.

CodeByTarek commented:
2025-07-30 22:35:25.622  6980-6980  AndroidRuntime          net.adhikary.mrtbuddy                E  FATAL EXCEPTION: main
                                                                                                    Process: net.adhikary.mrtbuddy, PID: 6980
                                                                                                    java.lang.IllegalArgumentException: Key "1749816000000_Dhaka University_Unknown Station (0)" was already used. If you are using LazyColumn/Row please make sure you provide a unique key for each item.
                                                                                                    	at androidx.compose.ui.layout.LayoutNodeSubcompositionsState.subcompose(SubcomposeLayout.kt:453)
                                                                                                    	at androidx.compose.ui.layout.LayoutNodeSubcompositionsState$Scope.subcompose(SubcomposeLayout.kt:895)
                                                                                                    	at androidx.compose.foundation.lazy.layout.LazyLayoutMeasureScopeImpl.measure-0kLqBqw(LazyLayoutMeasureScope.kt:125)
                                                                                                    	at androidx.compose.foundation.lazy.LazyListMeasuredItemProvider.getAndMeasure-0kLqBqw(LazyListMeasuredItemProvider.kt:55)
                                                                                                    	at androidx.compose.foundation.lazy.LazyListMeasuredItemProvider.getAndMeasure-0kLqBqw$default(LazyListMeasuredItemProvider.kt:49)
                                                                                                    	at androidx.compose.foundation.lazy.LazyListMeasureKt.measureLazyList-x0Ok8Vo(LazyListMeasure.kt:214)
                                                                                                    	at androidx.compose.foundation.lazy.LazyListKt$rememberLazyListMeasurePolicy$1$1.invoke-0kLqBqw(LazyList.kt:326)
                                                                                                    	at androidx.compose.foundation.lazy.LazyListKt$rememberLazyListMeasurePolicy$1$1.invoke(LazyList.kt:183)
                                                                                                    	at androidx.compose.foundation.lazy.layout.LazyLayoutKt$LazyLayout$3$2$1.invoke-0kLqBqw(LazyLayout.kt:119)
                                                                                                    	at androidx.compose.foundation.lazy.layout.LazyLayoutKt$LazyLayout$3$2$1.invoke(LazyLayout.kt:112)
                                                                                                    	at androidx.compose.ui.layout.LayoutNodeSubcompositionsState$createMeasurePolicy$1.measure-3p2s80s(SubcomposeLayout.kt:725)
                                                                                                    	at androidx.compose.ui.node.InnerNodeCoordinator.measure-BRTryo0(InnerNodeCoordinator.kt:135)
                                                                                                    	at androidx.compose.ui.graphics.SimpleGraphicsLayerModifier.measure-3p2s80s(GraphicsLayerModifier.kt:646)
                                                                                                    	at androidx.compose.ui.node.LayoutModifierNodeCoordinator.measure-BRTryo0(LayoutModifierNodeCoordinator.kt:188)
                                                                                                    	at androidx.compose.foundation.layout.FillNode.measure-3p2s80s(Size.kt:699)
                                                                                                    	at androidx.compose.ui.node.LayoutModifierNodeCoordinator.measure-BRTryo0(LayoutModifierNodeCoordinator.kt:188)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeLayoutDelegate$performMeasureBlock$1.invoke(LayoutNodeLayoutDelegate.kt:316)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeLayoutDelegate$performMeasureBlock$1.invoke(LayoutNodeLayoutDelegate.kt:315)
                                                                                                    	at androidx.compose.runtime.snapshots.Snapshot$Companion.observe(Snapshot.kt:503)
                                                                                                    	at androidx.compose.runtime.snapshots.SnapshotStateObserver$ObservedScopeMap.observe(SnapshotStateObserver.kt:502)
                                                                                                    	at androidx.compose.runtime.snapshots.SnapshotStateObserver.observeReads(SnapshotStateObserver.kt:258)
                                                                                                    	at androidx.compose.ui.node.OwnerSnapshotObserver.observeReads$ui_release(OwnerSnapshotObserver.kt:133)
                                                                                                    	at androidx.compose.ui.node.OwnerSnapshotObserver.observeMeasureSnapshotReads$ui_release(OwnerSnapshotObserver.kt:113)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeLayoutDelegate.performMeasure-BRTryo0(LayoutNodeLayoutDelegate.kt:1775)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeLayoutDelegate.access$performMeasure-BRTryo0(LayoutNodeLayoutDelegate.kt:40)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeLayoutDelegate$MeasurePassDelegate.remeasure-BRTryo0(LayoutNodeLayoutDelegate.kt:696)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeLayoutDelegate$MeasurePassDelegate.measure-BRTryo0(LayoutNodeLayoutDelegate.kt:672)
                                                                                                    	at androidx.compose.animation.AnimatedEnterExitMeasurePolicy.measure-3p2s80s(AnimatedVisibility.kt:812)
                                                                                                    	at androidx.compose.ui.node.InnerNodeCoordinator.measure-BRTryo0(InnerNodeCoordinator.kt:135)
                                                                                                    	at androidx.compose.animation.EnterExitTransitionModifierNode.measure-3p2s80s(EnterExitTransition.kt:1173)
                                                                                                    	at androidx.compose.ui.node.LayoutModifierNodeCoordinator.measure-BRTryo0(LayoutModifierNodeCoordinator.kt:188)
                                                                                                    	at androidx.compose.ui.graphics.BlockGraphicsLayerModifier.measure-3p2s80s(GraphicsLayerModifier.kt:578)
                                                                                                    	at androidx.compose.ui.node.LayoutModifierNodeCoordinator.measure-BRTryo0(LayoutModifierNodeCoordinator.kt:188)
                                                                                                    	at androidx.compose.animation.AnimatedVisibilityKt$AnimatedVisibilityImpl$1$1.invoke-3p2s80s(AnimatedVisibility.kt:699)
2025-07-30 22:35:25.624  6980-6980  AndroidRuntime          net.adhikary.mrtbuddy                E  	at androidx.compose.animation.AnimatedVisibilityKt$AnimatedVisibilityImpl$1$1.invoke(AnimatedVisibility.kt:698)
                                                                                                    	at androidx.compose.ui.layout.LayoutModifierImpl.measure-3p2s80s(LayoutModifier.kt:294)
                                                                                                    	at androidx.compose.ui.node.LayoutModifierNodeCoordinator.measure-BRTryo0(LayoutModifierNodeCoordinator.kt:188)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeLayoutDelegate$performMeasureBlock$1.invoke(LayoutNodeLayoutDelegate.kt:316)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeLayoutDelegate$performMeasureBlock$1.invoke(LayoutNodeLayoutDelegate.kt:315)
                                                                                                    	at androidx.compose.runtime.snapshots.Snapshot$Companion.observe(Snapshot.kt:2441)
                                                                                                    	at androidx.compose.runtime.snapshots.SnapshotStateObserver$ObservedScopeMap.observe(SnapshotStateObserver.kt:502)
                                                                                                    	at androidx.compose.runtime.snapshots.SnapshotStateObserver.observeReads(SnapshotStateObserver.kt:258)
                                                                                                    	at androidx.compose.ui.node.OwnerSnapshotObserver.observeReads$ui_release(OwnerSnapshotObserver.kt:133)
                                                                                                    	at androidx.compose.ui.node.OwnerSnapshotObserver.observeMeasureSnapshotReads$ui_release(OwnerSnapshotObserver.kt:113)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeLayoutDelegate.performMeasure-BRTryo0(LayoutNodeLayoutDelegate.kt:1775)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeLayoutDelegate.access$performMeasure-BRTryo0(LayoutNodeLayoutDelegate.kt:40)
                                                                                                    	at androidx.compose.ui.node.LayoutNodeLayoutDelegate$MeasurePassDelegate.remeasure-BRTryo0(LayoutNodeLayoutDelegate.kt:696)
                                                                                                    	at androidx.compose.ui.node.LayoutNode.remeasure-_Sx5XlM$ui_release(LayoutNode.kt:1221)
                                                                                                    	at androidx.compose.ui.node.LayoutNode.remeasure-_Sx5XlM$ui_release$default(LayoutNode.kt:1212)
                                                                                                    	at androidx.compose.ui.node.MeasureAndLayoutDelegate.doRemeasure-sdFAvZA(MeasureAndLayoutDelegate.kt:369)
                                                                                                    	at androidx.compose.ui.node.MeasureAndLayoutDelegate.remeasureAndRelayoutIfNeeded(MeasureAndLayoutDelegate.kt:566)
                                                                                                    	at androidx.compose.ui.node.MeasureAndLayoutDelegate.remeasureAndRelayoutIfNeeded$default(MeasureAndLayoutDelegate.kt:534)
                                                                                                    	at androidx.compose.ui.node.MeasureAndLayoutDelegate.measureAndLayout(MeasureAndLayoutDelegate.kt:390)
                                                                                                    	at androidx.compose.ui.platform.AndroidComposeView.measureAndLayout(AndroidComposeView.android.kt:1273)
                                                                                                    	at androidx.compose.ui.node.Owner.measureAndLayout$default(Owner.kt:248)
                                                                                                    	at androidx.compose.ui.platform.AndroidComposeView.dispatchDraw(AndroidComposeView.android.kt:1555)
                                                                                                    	at android.view.View.draw(View.java:23892)
                                                                                                    	at android.view.View.updateDisplayListIfDirty(View.java:22756)
                                                                                                    	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4540)
                                                                                                    	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4513)
                                                                                                    	at android.view.View.updateDisplayListIfDirty(View.java:22712)
                                                                                                    	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4540)
                                                                                                    	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4513)
                                                                                                    	at android.view.View.updateDisplayListIfDirty(View.java:22712)
                                                                                                    	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4540)
                                                                                                    	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4513)
                                                                                                    	at android.view.View.updateDisplayListIfDirty(View.java:22712)
                                                                                                    	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4540)
                                                                                                    	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4513)
                                                                                                    	at android.view.View.updateDisplayListIfDirty(View.java:22712)
                                                                                                    	at android.view.ThreadedRenderer.updateViewTreeDisplayList(ThreadedRenderer.java:694)
                                                                                                    	at android.view.ThreadedRenderer.updateRootDisplayList(ThreadedRenderer.java:700)
                                                                                                    	at android.view.ThreadedRenderer.draw(ThreadedRenderer.java:798)
                                                                                                    	at android.view.ViewRootImpl.draw(ViewRootImpl.java:4939)
                                                                                                    	at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:4643)
                                                                                                    	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:3822)
                                                                                                    	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:2465)
                                                                                                    	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:9305)
                                                                                                    	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1339)
                                                                                                    	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1348)
2025-07-30 22:35:25.624  6980-6980  AndroidRuntime          net.adhikary.mrtbuddy                E  	at android.view.Choreographer.doCallbacks(Choreographer.java:952)
                                                                                                    	at android.view.Choreographer.doFrame(Choreographer.java:882)
                                                                                                    	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1322)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:958)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:205)
                                                                                                    	at android.os.Looper.loop(Looper.java:294)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:8177)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
2025-07-30 22:35:25.634  6980-6980  Process                 net.adhikary.mrtbuddy                I  Sending signal. PID: 6980 SIG: 9
---------------------------- PROCESS ENDED (6980) for package net.adhikary.mrtbuddy ----------------------------

mojahid2021 committed:
feat(CardItem, FareCalculatorComponents, HistoryScreen, MainScreen): update UI components with improved shapes and padding

mojahid2021 committed:
feat(MoreScreen): redesign with modern UI components and enhanced user experience

mojahid2021 committed:
feat(CardItem): enhance UI with improved animations, dropdown menu, and status indicators

mojahid2021 committed:
feat(CardItem): enhance card gradient and UI transparency for improved visual appeal

mojahid2021 committed:
feat(CardItem, FareCalculatorScreen): update watermark typography and adjust content padding for improved layout

mojahid2021 committed:
feat(BalanceCard): update header height, adjust padding, and improve layout arrangement

mojahid2021 committed:
feat(BalanceCard, MainScreen): enhance layout with wrapContentHeight and improved spacing

mojahid2021 committed:
feat(build): update versionCode and versionName to 25 and 0.0.25

mojahid2021 committed:
feat(BalanceCard, MoreScreen): enhance layout and design with new icons, improved spacing, and updated text instructions

mojahid2021 committed:
feat(TransactionListScreen): enhance key generation for transaction items to ensure uniqueness

mojahid2021 commented:
@CodeByTarek Now, Check again and tell me if you still see the same problem.

CodeByTarek commented:
@mojahid2021 Good work. 😀 Let's fix the rest.

mojahid2021 committed:
1. feat(BalanceCard, FareCalculator): update text prompts and improve layout for better user guidance
2. MRT Pass Related issue fixed

mojahid2021 committed:
feat(build): update versionCode and versionName to 26 and 0.0.26

CodeByTarek commented:
@aniruddha-adhikary Bhai, have a look. I've tested the app, and so far, it's working nicely. If anyone can see it, it would be nice to have a new release. 😀

mojahid2021 committed:
feat:
1. enhance UI text and layout across multiple screens for improved user experience
2. Added Translation

CodeByTarek committed:
Updated Strings and made them sounds better and clear

CodeByTarek committed:
Minor change

CodeByTarek committed:
Minor changes

mojahid2021 committed:
Merge pull request #3 from CodeByTarek/master

Updated Strings and made them sounds better and clear

mojahid2021 committed:
feat(build): update versionCode and versionName to 27 and 0.0.27

CodeByTarek committed:
Update FareCalculatorComponents.kt

Minor Change for better visual

CodeByTarek committed:
Save local changes to FareCalculatorComponents.kt

CodeByTarek committed:
Merge branch 'master' of https://github.com/CodeByTarek/mrt-buddy

CodeByTarek committed:
Minor changes

CodeByTarek committed:
Minor changes for better UI

CodeByTarek committed:
Minor changes for better UI

mojahid2021 committed:
Merge pull request #4 from CodeByTarek/master

Update FareCalculatorComponents.kt

mojahid2021 committed:
feat(FareCalculator, HistoryScreen, MainScreen, StationService, TransactionListScreen): update UI text and improve localization for better user experience

mojahid2021 committed:
Merge pull request #5 from mojahid2021/version0.27

feat(FareCalculator, HistoryScreen, MainScreen, StationService, Trans…

mojahid2021 committed:
feat(Translation): added missing translations

mojahid2021 committed:
feat(Translation): add scan failed message and localization

CodeByTarek committed:
Refactored Divider to HorizontalDivider and minor updates

mojahid2021 committed:
Merge pull request #7 from CodeByTarek/master

Refactored Divider to HorizontalDivider and minor updates

mojahid2021 committed:
feat(FareCalculator): add station swapping functionality and improve UI components

mojahid2021 committed:
feat(FareCalculator): enhance UI components with improved accessibility and user guidance

mojahid2021 committed:
feat(FareCalculator): implement inline dropdown menu for station selection and enhance UI components

mojahid2021 committed:
fix(BalanceCard): adjust button width and improve text formatting for better UI consistency

mojahid2021 committed:
feat(FareCalculator): reorganize station selection layout for improved usability and clarity

mojahid2021 committed:
feat(FareCalculator): add animated swap functionality for station selection

mojahid2021 committed:
feat(FareCalculator): introduce ModernHeader component for improved header design and functionality

mojahid2021 committed:
feat(FareCalculator): enhance FareDisplayCard for responsive layout and improved usability

mojahid2021 committed:
refactor(FareCalculator): improve layout structure and readability in FareDisplayCard

mojahid2021 committed:
feat(FareCalculator): enhance FareDisplayCard layout with improved structure, balance summary, and user actions

mojahid2021 committed:
refactor(FareCalculator): inline ModernHeader into FareCalculatorComponents and remove TravelInfoCard

mojahid2021 committed:
refactor(FareCalculator): streamline header layout in FareCalculatorComponents and enhance visual consistency

mojahid2021 committed:
refactor(FareCalculator): replace CombinedRouteDropdown with inline dropdowns for improved usability

mojahid2021 committed:
refactor(FareCalculator): move swap button to header and simplify animation logic

mojahid2021 committed:
refactor(FareCalculator): enhance station selection UI with improved layout and spacing

mojahid2021 committed:
refactor(CardItem): replace clickable with pointerInput for improved gesture handling

MahabubKarim reviewed (commented):
Getting well structured day by say!

khaled-0 commented:
looks pretty -  YES
necessary - Probably Not

I personally would prefer Material3 Expressive in terms of redesign
